#519021910721-陆伊敏
## 下订单一共包含如下几个步骤：
+ orderInfoController接收前端发来的请求，通过placeOrder函数将订单请求加入待处理队列。
+ orderInfoServiceImpl中的placeOrder函数拿取队列中的订单请求，处理数据并调用dao层的函数处理。
+ 调用orderInfoDao中的placeOrder函数向数据库中插入新的订单。
+ 调用orderInfoDao中的addOrderItem函数向数据库中插入新的订单项。
+ 调用bookDao中的placeOrder函数修改书籍库存
+ 调用cartDao中的deleteCartItem函数删除用户购物车中的数据项。
+ 调用userDao修改用户消费总额

## 全部使用REQUIRED
required保证了所有函数都能在同一个事务中进行，orderInfoServiceImpl的placeOrder函数开启了一个新事务，后续所有操作都在这个事务中进行。如果某个步骤出错，则会整体进行回滚；如果没有出错，则会在placeOrder完全结束后提交事务。
（使用postman测试书籍购买量大于书籍库存的情况，测试结果表明事务会中断并回滚。）

## 全部使用MANDATORY
MANDATORY必须在一个已有的事务中执行，如果没有事务则抛出异常，因此下订单后，placeOrder函数不能正常执行，会抛出异常org.springframework.jms.listener.adapter.ListenerExecutionFailedException: Listener method 'public void com.reins.bookstore.serviceimpl.OrderInfoServiceImpl.placeOrder(com.reins.bookstore.entity.OrderInQueue)' threw exception; nested exception is org.springframework.transaction.IllegalTransactionStateException: No existing transaction found for transaction marked with propagation 'mandatory'

## service层的placeOrder函数使用REQUIRED，其余使用MANDATORY
下订单时，placeOrder函数会开启一个新事务，后续的函数虽然是MANDATORY，但是由于已经有了一个事务，所以可以正常执行。但是如果单独使用这些dao层的函数，仍旧会抛出异常。

## 全部用NEVER
NEVER需要不在事务中执行，由于没有任何一个被调用的方法会开启事务，所以一般情况下可以正常执行下订单功能。但是，如果遇到书籍库存小于书籍购买量等特殊情况，则会因为没有事务回滚功能而产生错误。

## service层的placeOrder函数使用REQUIRED，其余使用NEVER
NEVER需要不在事务中执行，但是placeOrder已经开启了事务，所以会抛出异常org.springframework.transaction.IllegalTransactionStateException: Existing transaction found for transaction marked with propagation 'never'

##全部使用REQUIRES_NEW
REQUIRES_NEW会创建一个新的事务，全部使用REQUIRES_NEW可以正常执行一般的下订单流程，但在遇到书籍库存小于书籍购买量等特殊情况时，由于每个被调用的方法都在一个新事务中执行，因此无法在某个方法出错时，对之前进行过的操作进行回滚，导致了错误的出现。
当然，REQUIRES_NEW适合用于一些不重要的方法（即便出错了也不会影响整个流程，相反，如果因为该方法出错而回滚整个事务，反而会导致效率降低）

